agent/kafka_handler.go

```diff
@@ -17,6 +33,136 @@ type kafkaHandler struct {
+	resp := &kafkaprotocol.CreateTopicsResponse{
+		ThrottleTimeMs: 0,
+		Topics:         make([]kafkaprotocol.CreateTopicsResponseCreatableTopicResult, len(req.Topics)),
+	}
+
+	for tidx, topic := range req.Topics {
+		retentionTime := defaultRetentionTime
+		respConfigs := make([]kafkaprotocol.CreateTopicsResponseCreatableTopicConfigs, len(topic.Configs))
+
+		// Parse custom retention time from topic configs if provided
+		for cidx, config := range topic.Configs {
+			if common.SafeDerefStringPtr(config.Name) == "retention.ms" {
+				retentionMs, err := strconv.Atoi(common.SafeDerefStringPtr(config.Value))
+				if err == nil {
+					retentionTime = time.Duration(retentionMs) * time.Millisecond
+				}
+			}
+
+			respConfigs[cidx] = kafkaprotocol.CreateTopicsResponseCreatableTopicConfigs{
+				Name:  config.Name,
+				Value: config.Value,
+			}
+		}
+
+		var errMsg string
+		errCode := int16(0)
+		derefTopicName := common.SafeDerefStringPtr(topic.Name)
+
+		err := detectInvalidTopic(derefTopicName)
+		if err != nil {
+			errMsg = fmt.Sprintf("Invalid topic: %s (Reason: %s)\n", derefTopicName, err.Error())
+			errCode = int16(kafkaprotocol.ErrorCodeInvalidTopicException)
+		} else {
+			acl, err := k.agent.controlClientCache.GetClient()
+			if err != nil {
+				errMsg = err.Error()
+				errCode = kafkaprotocol.ErrorCodeCoordinatorNotAvailable
+			} else {
+				topicInfo := topicmeta.TopicInfo{
+					Name:           derefTopicName,
+					PartitionCount: int(topic.NumPartitions),
+					RetentionTime:  retentionTime,
+				}
+
+				err = acl.CreateTopic(topicInfo)
+				if err != nil {
+					errMsg = err.Error()
+					if strings.Contains(errMsg, "already exists") {
+						errCode = kafkaprotocol.ErrorCodeTopicAlreadyExists
+					} else {
+						errCode = kafkaprotocol.ErrorCodeInvalidTopicException
+					}
+				}
+			}
+		}
+
+		resp.Topics[tidx] = kafkaprotocol.CreateTopicsResponseCreatableTopicResult{
+			Name:              topic.Name,
+			ErrorCode:         errCode,
+			ErrorMessage:      &errMsg,
+			NumPartitions:     topic.NumPartitions,
+			Configs:           respConfigs,
+		}
+	}
+	return completionFunc(resp)
+}
+
+func (k *kafkaHandler) HandleDeleteTopicsRequest(hdr *kafkaprotocol.RequestHeader, req *kafkaprotocol.DeleteTopicsRequest, completionFunc func(resp *kafkaprotocol.DeleteTopicsResponse) error) error {
+	resp := &kafkaprotocol.DeleteTopicsResponse{
+		ThrottleTimeMs: 0,
+		Responses:      make([]kafkaprotocol.DeleteTopicsResponseDeletableTopicResult, len(req.TopicNames)),
+	}
+
+	for tidx, topicName := range req.TopicNames {
+		var errMsg string
+		errCode := int16(0)
+
+		acl, err := k.agent.controlClientCache.GetClient()
+		if err != nil {
+			errMsg = err.Error()
+			errCode = kafkaprotocol.ErrorCodeCoordinatorNotAvailable
+		} else {
+			err = acl.DeleteTopic(common.SafeDerefStringPtr(topicName))
+			if err != nil {
+				errMsg = err.Error()
+				if strings.Contains(errMsg, "not exist") {
```

Comment:
Don't check the error message - check the exception type and extract the error code